### DISCUSSION TITLE
Proper performance benchmarking

### DISCUSSION THREAD

**Discussion by LemurPwned** [milvus-io/milvus#5379] [Q&A and General discussion] [answered]
Hi!
I'm currently trying to evaluate milvus in terms of recall.  My milvus is deployed in a K8s cluster.
I create 2 indexes on two separate collections -- a flat index and IVF_SQ8 index with nlist set to 1024. Then, I create 5M normalised random vectors and insert them into each collection. I flush and make sure the vectors are registered.

However, when I query both collections with the same query, I get exactly the same results (topK results are in the same order). This leads me to believe that the milvus uses exhaustive search in both cases despite me requesting the IVF_SQ8 index in the second collection.

Only when I submitted 50M vectors to each of the collections, I saw a difference in recall.
That leads me to a question -- was the index used when I had 5M vectors? Maybe I expected a difference when there shouldn't be one? Or, there's inbuilt threshold that decides to build the index when a specified number of vectors lands in the collection?

**Comment by shiyu22** [comment, marked as answer]
Hi,

The parameter related to index building is `index_file_size`, you set `index_file_size` when creating a collection from a client. This parameter specifies the size of each segment, and its default value is 1024 in MB. When the size of newly inserted vectors reaches the specified volume, Milvus packs these vectors into a new segment. In other words, newly inserted vectors do not go into a segment until they grow to the specified volume. When it comes to creating indexes, Milvus creates one index file for each segment. When conducting a vector search, Milvus searches all index files one by one.

So for 5M data, what is the dimension? If the inserted data size does not reach `index_file_size`, no index will be created. 

On the other hand, the query result also depends on the distribution space of your generated vector data, if the distribution is relatively uniform, indeed the query result will be strange.

**Reply by LemurPwned** [reply]
Thanks! This makes sense now -- the vector dimension was indeed relatively small ~50 (so that'd probably hit lower than 50M but much higher than 5M). I'll check setting smaller index size then :)

**Reply by shiyu22** [reply]
And you can also call `create_index()` manually when the amount of data does not reach the size of `index_file_size`. After creating the index, you can use SQL to check the status of the segment, please refer to this article:: https://blog.milvus.io/milvus-metadata-management-2-fields-in-the-metadata-table-3bf0d296ca6d

**Reply by LemurPwned** [reply]
Hmm that's interesting -- I've been creating index using pymilvus and `milvus_client.create_index(
    collection_name=ivf_sq8_collection_eval, index_type=IndexType.IVF_SQ8, params=indx_param)`. Do you mean calling it again on the same collection?

**Reply by shiyu22** [reply]
Yes, you can call it again. Otherwise, the index will be created automatically when the `index_file_size` is reached.

**Reply by LemurPwned** [reply]
Thanks! That's an interesting idea. By the way, when I create the index and then insert loads of vectors -- is it sufficient to see index name in the collection stats? In other words, how can I make sure programatically that the index has been fully built?

**Reply by shiyu22** [reply]
It can get the status with get_collection_stats(), you can refer to this #5234

**Reply by yhmo** [reply]
In version <= v1.0.0, the get_collection_stats() has a bug [#5050](https://github.com/milvus-io/milvus/issues/5050) that always return collection's index type for segments. We have fixed it in v1.1.0. After the fix, if a segment has not finished index, its index type will be "IDMAP".
